OpenRA.Mods.Common/Traits/BotModules/PowerDownBotModule.cs

```diff
@@ -0,0 +1,154 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Manages AI powerdown.")]
+	public class PowerDownBotModuleInfo : ConditionalTraitInfo
+	{
+		[Desc("Delay (in ticks) between toggling powerdown.")]
+		public readonly int Interval = 150;
+
+		[Desc("Order string that used for powerdown.")]
+		public readonly string OrderName = "PowerDown";
+
+		public override object Create(ActorInitializer init) { return new PowerDownBotModule(init.Self, this); }
+	}
+
+	public class PowerDownBotModule : ConditionalTrait<PowerDownBotModuleInfo>, IBotTick
+	{
+		readonly World world;
+		readonly Player player;
+		PowerManager playerPower;
+		int toggleTick;
+		readonly Func<Actor, bool> isToggledBuildingsValid;
```

Comment:
It would be better to make this `Func<BuildingPowerWrapper, bool>` and remove the `GetTogglePowerChanging` call.

Reply:
But I guess it will make code more complex here, meanwhile `GetTogglePowerChanging` is a function used in other place in this module.

Reply:
I will just drop `GetTogglePowerChanging` call here. I feel it senseless even for the mod that will have crazy idea on power mutiplier.